File: microsoft/medium/KthLargestElementInAnArray.java
class Solution {
    public int findKthLargest(int[] nums, int k) {
        // THOUGHT PROCESS:
        // Brute force: Sort array and return element at index n - k - O(n log n) time
        // This works but sorting the entire array is overkill
        //
        // Better: Min-heap of size k
        // Time: O(n log k). Space: O(k)

        // PSEUDOCODE:
        // 1. Maintain a min-heap of size k
        // 2. For each number, add to heap
        // 3. If heap size exceeds k, remove smallest element
        // 4. Root of heap is the kth largest element
        PriorityQueue<Integer> heap = new PriorityQueue<>();
        
        for (int num : nums) {
            // add current number to heap
            heap.add(num);
            // if heap exceeds size k, remove smallest element
            if (heap.size() > k) {
                heap.remove();
            }
        }

        // root contains kth largest element
        return heap.peek();
    }
}

/* BRUTE FORCE SOLUTION:
public int findKthLargest(int[] nums, int k) {
    Arrays.sort(nums);
    return nums[nums.length - k];
}
*/

File: meta/leetcode/kth_largest_element_in_an_array.py
import heapq
from typing import List

# Time: O(N log k) - we push/pop N elements in a heap of size k
# Space: O(k) - heap of size k
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        # use a min-heap to keep track of the k largest elements
        min_heap = []
        
        for num in nums:
            heapq.heappush(min_heap, num)
            # if heap exceeds size k, remove smallest element
            if len(min_heap) > k:
                heapq.heappop(min_heap)
        
        # the root of the min-heap is the kth largest element
        return min_heap[0]